Converter refactor

michael opened the pull request.

obuchtala committed:
Total refactor of html im-/exporter.

- simplified API for node converters
- node converters can reside in their own file
- using Component.$$ for export
- introduced a DOMElement abstraction to hide jQuery dependency

obuchtala committed:
Added test for Paragraph html import.

obuchtala committed:
More HtmlImporter tests and bugfixes.

obuchtala committed:
Fixes in html converter.

obuchtala committed:
Using QUnit server side.

Commit:
Fix typo in _list.scss.

Commit:
Linter and wording.

Commit:
Add hasClass method to DOMElement.

Commit:
Fix regression in HtmlImporter.

michael commented:
I've done an iteration on the demos (in future examples), covering the new converter API:

It's looking good, the API looks really clean and we have a clear separation now between Article definition, converter the editing app world.

@oliver---- can you review:

https://github.com/substance/demos/tree/beta2/note
https://github.com/substance/demos/tree/beta2/converter
https://github.com/substance/demos/tree/beta2/notepad

From my side we can merge the converter refactor. I will adjust also the Lens converter, but this could also be done after the Beta 2 release. 

#235 we may want to address before though?

Edit: #235 is done

obuchtala commented:
@michael Thx I am just starting the next iteration

obuchtala commented:
We can address #235 in an extra PR.

Examples looking good so far.

obuchtala commented:
We should wait until we have merged #234.

Commit:
Fix html converters for Blockquote and Codeblock.

obuchtala commented:
I will make this PR independent of #234.

obuchtala commented:
@michael don't push to this branch for now
I am rebasing it onto master and creating a new PR

obuchtala commented:
I finish the exporter things before we can merge.

obuchtala committed:
Make linter happy again.

obuchtala commented:
Need to fix ClipboardImporter and ClipboardExporter too

obuchtala committed:
Renamed HtmlImporter -> HTMLImporter, HtmlExporter -> HTMLExporter.

Moved implementation into parent classes DOMImporter and DOMExporter.

obuchtala committed:
Minor changes.

obuchtala committed:
Introduced ui/DOMElement.

Started to implement this interface in ui/DefaultDOMElement
and ui/VirtualDOMElement (WIP).

Commit:
Update DOMElement.js

obuchtala committed:
Implementing ui/DOMElement in ui/Component (WIP).

obuchtala committed:
Stub fix for ClipboardImporter/Exporter.

obuchtala committed:
Fix DOMElement implementations.

obuchtala committed:
First HTMLExporter test.

obuchtala committed:
Fix type tag matcher.

obuchtala committed:
Fix regressions.

obuchtala committed:
Added more HTMLExporter tests.

obuchtala committed:
Fix regressions.

obuchtala commented:
So! first iteration done.
Still I am using jQuery internally, but it has gone from the interface.
Later on I might try to get rid of it completely.

@michael u can find a test for HTMLExporter to see how the exporting spirit is.
Plz let me know if you think this could be improved.
And merge if you are happy so far.

obuchtala commented:
I will address other converter related stuff, such as Clipboard in another PR.

michael commented:
No rush on eliminating jQuery as long as we don't use it directly. I will utilize the exporter API in the note converter example now.

michael commented on packages/embed/EmbedHtmlConverter.js:
```diff
@@ -0,0 +1,21 @@
+      .html(node.html);
+  }
+
+};
```
A minor thing, but do if we want to be strict that filenames of classes start with a capital letter, we should call it `embedHtmlConverter.js` since we return an object here, not a full class. Or is that too picky?

obuchtala replied:
I agree.

michael commented:
Worked just fine: https://github.com/substance/demos/commit/cd60ee8378b16286c356c56852a0cd8cb3f993ff

michael commented:
Want to upgrade the Lens Importer so we can check how this plays with XML. We want to have XmlImporter XmlExporter entry points right? However we should get this merged now and work on the polishing in separate PR's.

obuchtala commented:
Did we go for XML now in the science blogger?

obuchtala commented:
Beyond that, having XML converters for all of our nodes would be fine.

michael commented:
Lens Writer is already XML yes. But abusing the old HtmlImporter API for importing that.

For most nodes XML converter would be the same as HTML importer (e.g. paragraph). should we introduce redundancy there? Maybe we could call the generic converters just DOMConverters?

obuchtala commented:
I would rather add a delegator. Then there is no confusion.

obuchtala commented:
@michael I would need a lens version working with latest master. Is it ok if you address that LensConverter update next?

michael commented:
Will do that tomorrow!
On Mo., 16. Nov. 2015 at 19:25, Oliver Buchtala notifications@github.com
wrote:

> @michael https://github.com/michael I would need a lens version working
> with latest master. Is it ok if you address that LensConverter update next?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/substance/substance/pull/242#issuecomment-157127552.

obuchtala commented:
Cool :)
